server/lib/nl/fulfillment/filter_with_dual_vars.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import logging
+from typing import List
+
+from server.lib.nl import utils
+from server.lib.nl.detection import ContainedInPlaceType
+from server.lib.nl.detection import Place
+from server.lib.nl.fulfillment.base import add_chart_to_utterance
+from server.lib.nl.fulfillment.base import ChartVars
+from server.lib.nl.fulfillment.base import open_top_topics_ordered
+from server.lib.nl.fulfillment.base import populate_charts
+from server.lib.nl.fulfillment.base import PopulateState
+from server.lib.nl.utterance import ChartOriginType
+from server.lib.nl.utterance import ChartType
+from server.lib.nl.utterance import Utterance
+
+# TODO: Support per-capita
+# TODO: Increase this after frontend handles comparedPlaces better.
+_MAX_PLACES_TO_RETURN = 7
+
+
+def _get_dual_sv_part(uttr: Utterance, idx: int) -> List[str]:
+  for c in uttr.multi_svs.candidates:
+    if len(c.parts) == 2:
+      return c.parts[idx].svs
+  return []
+
+
+#
+# Computes ranked list of places after applying the filter and
+# retrieve a var for those places.
+#
+def populate(uttr: Utterance):
+  quantity = utils.get_quantity(uttr)
+  if not quantity:
+    uttr.counters.err('filter-with-dual-var_missingquantity', 1)
+    return False
+  place_type = utils.get_contained_in_type(uttr)
+  if not place_type:
+    place_type = ContainedInPlaceType.DEFAULT_TYPE
+
+  # We always assume the 2nd SV is the filter SV.
+  svs = _get_dual_sv_part(uttr, 1)
+  if not svs:
+    uttr.counters.err('filter-with-dual-var_notdualsv', 1)
+    return False
+
+  # Update `svs` in uttr across the call to populate_charts()
+  orig_svs = uttr.svs
+  uttr.svs = svs
+  found = populate_charts(
+      PopulateState(uttr=uttr,
+                    main_cb=_populate_cb,
+                    place_type=place_type,
+                    quantity=quantity))
+  uttr.svs = orig_svs
+  return found
+
+
+def _populate_cb(state: PopulateState, chart_vars: ChartVars,
```

Comment:
nit: The _populate_cb() function here has overlap with the similar function in the filter_with_single_var.py file. Is there room to move things in to utils etc?

Reply:
Good point.  Added TODO.